src/Modules/OrchardCore.Commerce/Services/ProductService.cs

```diff
@@ -79,14 +79,14 @@ private static void FillField(ContentPart part, ContentPartFieldDefinition field
 
     public string GetVariantKey(string sku)
     {
-        var dashIndex = sku.IndexOf(value: "-", StringComparison.InvariantCulture);
-        return dashIndex == -1 ? sku : sku[(dashIndex + 1)..];
+        var dashIndex = ProductServiceExtensions.GetDashIndex(sku);
+        return dashIndex == -1 ? sku : sku[(dashIndex + 1)..].ToUpperInvariant();
```

Comment:
This is still wrong. Let me clarify, this function is okay (apart from `.ToUpperInvariant()` see below), I just commented here, because there is no specific place where I can comment this.

`productPart.Sku + '-' + value.Classify().ToUpperInvariant` is still missing. It's okay that we can give this function an SKU and a variant key and it will get the variant key, but the variant key is generated nowhere. There is no function to currently get the SKU + variant key. 

Like if we have a size L, black shirt, and its base SKU is SHIRT, there is no way to get SHIRT-L-BLACK (so the SKU and  the variant key). 

Dávid mentioned this here (please read this carefully, also see the linked code): https://github.com/OrchardCMS/OrchardCore.Commerce/issues/135#issuecomment-1251175671

You don't need to save it, but there should be a way to generate it. It might not be used currently, but it should be implemented. (I think you also asked this:

> Makes sense; just to clarify, these display surfaces are to be implemented in the future and do not yet exist, right?

)
Though please check if anything need to be replaced by the SKU + variant key combination where `PriceVariantPart` is used for a product.

Also, you don't need to convert it to `ToUpperVariant()`, the base SKU will be uppercase anyway and if you create the above-mentioned variant key generator function you can attach the variant key there as uppercase, so it will "arrive" here as uppercase already.

@DAud-IcI Could you please confirm that I'm right? Since the idea was yours.

Reply:
The problem is that currently there is no relation between the SKU fragment and the `PriceVariantsPart.Variants` dictionary key. The goal is to be able to look up the specific variant price from the variant SKU. The best approach would be to update [here](https://github.com/OrchardCMS/OrchardCore.Commerce/blob/2ed945f5996a71cba81b7b17ab61e1f86713ba53/src/Modules/OrchardCore.Commerce/Drivers/PriceVariantsPartDisplayDriver.cs#L81) and [here](https://github.com/OrchardCMS/OrchardCore.Commerce/blob/6645257c7dd248e30ee915c6527524bb4d657131/src/Modules/OrchardCore.Commerce/Services/PriceVariantProvider.cs#L65), basically uppercase and classify the dictionary key to get the SKU style we want. Then we won't have to convert back and forth between two different key types.

Once updated, please extract [this section](https://github.com/OrchardCMS/OrchardCore.Commerce/blob/6645257c7dd248e30ee915c6527524bb4d657131/src/Modules/OrchardCore.Commerce/Services/PriceVariantProvider.cs#L60-L69) into a method of `ShoppingCartItem` because it will be reused by future features. Also create a unit test to verify that the SKU variant key is generated correctly from the `ShoppingCartItem` attributes.

I agree that you don't need `ToUpperInvariant()`, but on the other hand please add `StringComparer.OrdinalIgnoreCase` to the Dictionary constructor [here](https://github.com/OrchardCMS/OrchardCore.Commerce/blob/a4408afb203b0e5a135a19c38f18fac2ab0f4efe/src/Modules/OrchardCore.Commerce/Models/PriceVariantsPart.cs#L12) to avoid backwards compatibility issues with existing instances.

Reply:
After adding uppercase conversion and classification to the highlighted parts in **PriceVariantsPartDisplayDriver** and **PriceVariantProvider**, it's not clear to me if a standalone SKU-generator method is necessary?

Primary reason I've been slow to address this part is I've still been struggling to understand how exactly it's supposed to work. Please let me know if my latest additions are correct.

Reply:
It's okay. I wanted a standalone method for future-proofing, but we can do that when the need actually arises.